src/clients/python/python_bindings.zig

```diff
@@ -0,0 +1,502 @@
+        \\tb_client_init.argtypes = [ctypes.POINTER(Client), c_uint128, ctypes.c_char_p,
+        \\                           ctypes.c_uint32, ctypes.c_void_p, OnCompletion]
+        \\
+        \\# Initialize a new TigerBeetle client which echos back any data submitted.
+        \\tb_client_init_echo = tbclient.tb_client_init_echo
+        \\tb_client_init_echo.restype = Status
+        \\tb_client_init.argtypes = [ctypes.POINTER(Client), c_uint128, ctypes.c_char_p,
+        \\                           ctypes.c_uint32, ctypes.c_void_p, OnCompletion]
+        \\
+        \\# Closes the client, causing any previously submitted packets to be completed with
+        \\# `TB_PACKET_CLIENT_SHUTDOWN` before freeing any allocated client resources from init.
+        \\# It is undefined behavior to use any functions on the client once deinit is called.
+        \\tb_client_deinit = tbclient.tb_client_deinit
+        \\tb_client_deinit.restype = None
+        \\tb_client_deinit.argtypes = [Client]
+        \\
+        \\# Submit a packet with its operation, data, and data_size fields set.
+        \\# Once completed, `on_completion` will be invoked with `on_completion_ctx` and the given
+        \\# packet on the `tb_client` thread (separate from caller's thread).
+        \\tb_client_submit = tbclient.tb_client_submit
+        \\tb_client_submit.restype = None
+        \\tb_client_submit.argtypes = [Client, ctypes.POINTER(CPacket)]
+        \\
+    , .{});
+
+    inline for (.{ true, false }) |is_async| {
+        const prefix_class = if (is_async) "Async" else "";
+        const prefix_fn = if (is_async) "async " else "";
+        const prefix_call = if (is_async) "await " else "";
+
+        buffer.print(
+            \\class {s}StateMachineMixin:
+            \\    _submit: Callable[[Operation, Any, Any, Any], Any]
+            \\
+        , .{prefix_class});
+
+        inline for (std.meta.fields(StateMachine.Operation)) |operation| {
+            const op: StateMachine.Operation = @enumFromInt(operation.value);
+
+            const event_type = if (StateMachine.event_is_slice(op))
+                "list[" ++ comptime resolve_type_dataclass(StateMachine.EventType(op)) ++ "]"
+            else
+                comptime resolve_type_dataclass(StateMachine.EventType(op));
+
+            const result_type = "list[" ++
+                comptime resolve_type_dataclass(StateMachine.ResultType(op)) ++
+                "]";
+
+            // For ergonomics, the client allows calling things like .query_accounts(filter) even
+            // though the _submit function requires a list for everything. Wrap them here.
+            const event_name_or_list = comptime if (!StateMachine.event_is_slice(op))
+                "[" ++ StateMachine.event_name(op) ++ "]"
+            else
+                StateMachine.event_name(op);
+
+            // TODO: Pulse shouldn't be hardcoded.
+            if (op != .pulse) {
+                buffer.print(
```

Comment:
This block is pretty dense and almost unreadable here. Between the high levels of indentation and parameters. Still trying to figure out something nicer.

Reply:
When scaling up, the best way to write such kinds of things I've found is to move the entire template _up_ and let the logic follow template, like here:

https://github.com/tigerbeetle/blog/commit/279d1d83912a5d0393c0014b34b24236eb01a4bc#diff-f2948449439cc27583df5f215cf1af9256df2de90031f03f3d2e0dc63cfcd4eeR138

Not sure if its worth inventing some DSL for python though!

The good news, it doesn't _really_ matter if this code is readable or not, the important thing to review is the generated code!

Reply:
This could be marginally better:

```
fn emit_method(buffer: *Buffer, comptime operation: std.builtin.Type.EnumField, options: struct {
    is_async: bool,
}) void {
    const op: StateMachine.Operation = @enumFromInt(operation.value);

    const event_type = comptime if (StateMachine.event_is_slice(op))
        "list[" ++ resolve_type_dataclass(StateMachine.EventType(op)) ++ "]"
    else
        resolve_type_dataclass(StateMachine.EventType(op));

    const result_type =
        comptime "list[" ++ resolve_type_dataclass(StateMachine.ResultType(op)) ++ "]";

    // For ergonomics, the client allows calling things like .query_accounts(filter) even
    // though the _submit function requires a list for everything. Wrap them here.
    const event_name_or_list = comptime if (!StateMachine.event_is_slice(op))
        "[" ++ StateMachine.event_name(op) ++ "]"
    else
        StateMachine.event_name(op);

    buffer.print(
        \\    {[prefix_fn]s}def {[fn_name]s}(self, {[event_name]s}: {[event_type]s}) -> {[result_type]s}:
        \\        return {[prefix_call]s}self._submit(
        \\            Operation.{[uppercase_name]s},
        \\            {[event_name_or_list]s},
        \\            {[event_type_c]s},
        \\            {[result_type_c]s},
        \\        )
    ,
        .{
            .prefix_fn = if (options.is_async) "async " else "",
            .fn_name = operation.name,
            .event_name = StateMachine.event_name(op),
            .event_type = event_type,
            .result_type = result_type,
            .event_name_or_list = event_name_or_list,
            .prefix_call = if (options.is_async) "await " else "",
            .uppercase_name = to_uppercase(operation.name),
            .event_type_c = ctype_type_name(StateMachine.EventType(op)),
            .result_type_c = ctype_type_name(StateMachine.ResultType(op)),
        },
    );
}
````

* extracting stuff into a function gives you couple of more levels of indentation to work with
* for arguments, they don't fit on the single line
* with `\\` literal, that means that the output would also be split
* but the length of the output should be _roughly_ the same as the length of the input
* and, indeed, with the existing impl some of the lines are longer than 100! So I think it's actually good to to argument per line for the output as well
* and tidy is smart, it checks that what's inside a `\\` is limited by 100, it doesn't count indentation in this case.